libraries/chain/controller.cpp

```diff
@@ -4066,6 +4066,15 @@ struct controller_impl {
                   "QC is_strong (${s1}) in block extension does not match is_strong_qc (${s2}) in header extension. Block number: ${b}",
                   ("s1", qc_proof.is_strong())("s2", new_qc_claim.is_strong_qc)("b", block_num) );
 
+      if (prev.valid) {
+         // compute finality mroot using previous block state and new qc claim
+         auto computed_finality_mroot = prev.get_finality_mroot_claim(new_qc_claim);
+         auto supplied_action_mroot = b->action_mroot;
+         EOS_ASSERT( computed_finality_mroot == supplied_action_mroot, block_validate_exception,
+                     "computed finality mroot (${computed}) does not match supplied action mroot ${supplied} by header extension. Block number: ${b}",
```

Comment:
Should include block id as well.

Reply:
This https://github.com/AntelopeIO/spring/blob/5247431b84f2c6d712494e650d71d3169fe04667/libraries/chain/controller.cpp#L4200 checks block id.

Reply:
Sorry, I meant include the block id in the error message.